### ISSUE TITLE
Setup RabbitMQ Worker for Job Submission

### ISSUE THREAD

**Issue by TroJan** [Cloud-CV/EvalAI#318]
A python worker(as of now) to get submission from queue, execute it with the evaluation script of the challenge(challenge id will be passed in the message) and do further tasks like sending mail, saving stdout and stderr outputs.

When a user makes a submission for a challenge, then a api to save his submission detail(like time at which the submission is made, participant team id, etc) will be requested. This api in turn will publish  a message to the queue. Assume that this message has key `submission.*.*`. This is actually the left part of the RabbitMq exchange

At the right side of the exchange, there will be python worker(will be definitely more than 1). These workers will listen on key `submission.*.*` on the exchange. Upon receiving a new message, the worker will execute the evaluation script of the desired challenge with the arguments passed in the message. stdout and stderr output will be captured and db updates, mail etc such tasks will be performed afterwards.

As discussed with @deshraj , we came up with two approaches . Listing down both approaches as separate comments

NOTE :: 
* Message key values may change

**Comment by TroJan** [comment]
Approach 1:

when the worker starts for the first time, fetch all the active challenges and its corresponding uploaded evaluation script. Dynamically require this evaluation script in a global variable named `EVALUATION_SCRIPTS`. So whenever any request comes for submission, the worker will fetch challenge id from that message and then simply call

EVALUATION_SCRIPTS[challenge_id].main(args)

This approach will work fine if at once all the active challenges are uploaded(this wont happen :P)
So the problem that comes is how this worker will be made aware of any new active challenge being added. One way to do this can be we can write a small worker which will poll the db at every X (6 hrs ?) interval of time and then publish a message with key `newchallenge.*.*`. This message when received at the exchange wil be delivered to each worker's queue.

So in a nutshell this approach brings

* 1 worker listening on 2 keys `newchallenge.*.*` and `submission.*.*`. 
* 1 worker polling db to fetch new active challenges.


__Problem/ Not Clear ??__

How will we redirect stdout and stderr when calling the evaluation script for a particular challenge.

**Comment by TroJan** [comment]
Approach 2:

Worker only listens on a single key `submission.*.*`. Upon receiving a message of this type, it extracts the challenge id from the message, then queries the db to url of the evaluation script. Fetches script from there and then use subprocess to call the script.

__Problem__

For each submission need to query database, query can be definitely cached, but still a headache.

**Comment by TroJan** [comment]
@deshraj @dexter1691 : your input please

**Comment by dexter1691** [comment]
@TroJan : this issue  is solved, right? 
Should we close?

**Comment by deshraj** [comment]
The first version of worker is ready. Closing it now.
